### ISSUE TITLE
TFLite Concatenation Fails on GPU delegate

### ISSUE THREAD

**Issue by mtngld** [tensorflow/tensorflow#38701]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes.
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Mac
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: Tested on Samsung Galaxy
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v2.1.0-rc2-17-ge5bf8de410 2.1.0
- Python version: 3.7
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

**Describe the current behavior**
Concatenation Op for non BCHW dimensions fails on GPU delegate, same model runs fine on CPU delegate.

(See below code for creating `model.tflite`)

The following benchmark runs successfully (CPU):
```
adb shell /data/local/tmp/benchmark_model      --graph=/data/local/tmp/model.tflite     --num_threads=4 --use_gpu=false --num_runs=1000
```

The following benchmark run fails (GPU):
```
adb shell /data/local/tmp/benchmark_model      --graph=/data/local/tmp/model.tflite     --num_threads=4 --use_gpu=true --num_runs=1000
```
Error message:
```
INFO: Initialized TensorFlow Lite runtime.
INFO: Created TensorFlow Lite delegate for GPU.
ERROR: TfLiteGpuDelegate Init: CONCATENATION: Dimensions are not BHWC: 2 1 30
ERROR: TfLiteGpuDelegate Prepare: delegate is not initialized
ERROR: Node number 1 (TfLiteGpuDelegateV2) failed to prepare.
```

**Describe the expected behavior**

TFLite Model should run concat op succesfully both on CPU and GPU.

**Standalone code to reproduce the issue**
```
import numpy as np
import tensorflow as tf

# Construct a basic model.
root = tf.train.Checkpoint()
root.v1 = tf.Variable(np.ones((1, 10), dtype=np.float32))
root.v2 = tf.Variable(np.ones((1, 20), dtype=np.float32))
root.f = tf.function(lambda x: tf.concat([root.v1, root.v2, x], axis=1))

# Save the model.
export_dir = "/tmp/test_saved_model"
input_data = tf.constant(np.ones((1, 30), dtype=np.float32))
to_save = root.f.get_concrete_function(input_data)
tf.saved_model.save(root, export_dir, to_save)

# Convert the model.
converter = tf.lite.TFLiteConverter.from_saved_model(export_dir)
tflite_model = converter.convert()
open("model.tflite", "wb").write(tflite_model)
```

**Comment by thaink** [comment]
The gpu delegate is not really friendly with non-constant tensors. Could you tried to use constant tensors in your model, convert it and try again?

**Comment by mtngld** [comment]
Do you mean changing `tf.Variable` to `tf.constant`?

Tried that, same issue occurs.

**Comment by thaink** [comment]
@mtngld You can use tf.Variable but set trainable=False

**Comment by mtngld** [comment]
Still fails with the same error

```
import numpy as np
import tensorflow as tf

shape = (1, 10)

# Construct a basic model.
root = tf.train.Checkpoint()
root.v1 = tf.Variable(np.ones(shape, dtype=np.float32), trainable=False)
root.v2 = tf.Variable(np.ones(shape, dtype=np.float32), trainable=False)
root.f = tf.function(lambda x: tf.concat([root.v1, root.v2, x], axis=1))

# Save the model.
export_dir = "/tmp/test_saved_model"
input_data = tf.constant(np.ones(shape, dtype=np.float32))
to_save = root.f.get_concrete_function(input_data)
tf.saved_model.save(root, export_dir, to_save)

# Convert the model.
converter = tf.lite.TFLiteConverter.from_saved_model(export_dir)
tflite_model = converter.convert()
open("model.tflite", "wb").write(tflite_model)
```

Output:
```
INFO: Initialized TensorFlow Lite runtime.
INFO: Created TensorFlow Lite delegate for GPU.
ERROR: TfLiteGpuDelegate Init: CONCATENATION: Dimensions are not BHWC
INFO: Created 0 GPU delegate kernels.
ERROR: TfLiteGpuDelegate Prepare: delegate is not initialized
ERROR: Node number 1 (TfLiteGpuDelegateV2) failed to prepare.

ERROR: Restored original execution plan after delegate application failure.
Failed to apply GPU delegate.
Benchmarking failed.
```

BTW, even when changing shape to `shape = (1, 1, 1, 10)` in order to match the BHWC (and concating on `axis=3`) requirement the benchmark fails on GPU with a different error:

```
Loaded model /data/local/tmp/model.tflite
INFO: Initialized TensorFlow Lite runtime.
INFO: Created TensorFlow Lite delegate for GPU.
ERROR: No selector for const
ERROR: Falling back to OpenGL
INFO: Initialized OpenGL-based API.
ERROR: TfLiteGpuDelegate Init: Suitable node shader is not found:
INFO: Created 0 GPU delegate kernels.
ERROR: TfLiteGpuDelegate Prepare: delegate is not initialized
ERROR: Node number 1 (TfLiteGpuDelegateV2) failed to prepare.

ERROR: Restored original execution plan after delegate application failure.
Failed to apply GPU delegate.
Benchmarking failed.
```

**Comment by thaink** [comment]
@impjdi Do you think this is a bug?
No suitable shader found for concat and no error message explain about that.

**Comment by impjdi** [comment]
ERROR: TfLiteGpuDelegate Init: CONCATENATION: Dimensions are not BHWC

I think that's the reason?

**Comment by thaink** [comment]
@impjdi I am mentioning about the later part. When BHWC was being used but there was another error.

**Comment by mtngld** [comment]
@thaink, @impjdi - thank you for taking a look at this issue.

Just to summarize - 

Looks like there are 2 different issues here:
1. BHWC concat op using GPU delegate -> `Suitable node shader is not found`
2. Non-BHWC concat op using GPU delegate -> `ERROR: TfLiteGpuDelegate Init: CONCATENATION: Dimensions are not BHWC`

(Note: Both operations work fine on CPU delegate)

**Comment by impjdi** [comment]
Thanks for elaborating.  The first thing happens here`//tensorflow/lite/delegates/gpu/gl/kernels/registry.cc` Could you please add a print statement and tell me which key is causing the issue?

**Comment by mtngld** [comment]
Sure,

I have added the following print 

```
    std::vector<std::string> errors;
    std::cout << "ctx.node->operation.type: " <<  ctx.node->operation.type << "\n";
    auto it = shaders_.find(ctx.node->operation.type);
```

Result - 
```
ctx.node->operation.type: const
```

**Comment by impjdi** [comment]
That's quite interesting.  To me it feels like it should have failed way before getting to this point for const not being a supported op.  Not sure why this slips through.

If you change tf constant to tf placeholder, does it work?

**Comment by mtngld** [comment]
I am using TF2.1 and TF2.2-RC3, I thought placeholders and `sess.run` are deprecated, not sure how to change the `tf.constant` into `tf.placeholder` and export the model.

**Comment by impjdi** [comment]
Hm, not a proficient TF user myself and can't give you guidance there =/

In the meantime, I found why it's letting `const` through.  For whatever reason, it's registered as an available op.  I created a change to make it clearer in the log (under review).

**Comment by mtngld** [comment]
Hi @impjdi - has the logging change that you mentioned passed the review and merged into master?

**Comment by impjdi** [comment]
Sorry, that got some comments and I never revisited.  Will take another look.

**Comment by mtngld** [comment]
Hi, any update on this issue? 

Just tested exporting in both 2.3.1 and tf-nightly (2.4.0-dev20200930), running on device with the [pre-built benchmark app](https://storage.googleapis.com/tensorflow-nightly-public/prod/tensorflow/release/lite/tools/nightly/latest/android_aarch64_benchmark_model.apk) - still seeing the same issue.

Error message is a bit more clear 

```
TfLiteGpuDelegate Init: CONCATENATION: Expected a 4D tensor of shape BxHxWxC but got 1x10
```

**Comment by mtngld** [comment]
Also,

The even when using BHWC, the shader error is still there:

```
TfLiteGpuDelegate Init: Suitable node shader is not found:
```

So anyway, both issues from https://github.com/tensorflow/tensorflow/issues/38701#issuecomment-620398307 are still there in tf-nightly

**Comment by impjdi** [comment]
Ah, the logging change dropped under the reviewer's radar, and I got the approval like a month ago, but then it got buried deep inside my list of changes, that I didn't notice the approval.  I saw that change last week, but saw that the approval was conditional and required some changes which I never got the time to do.  Let me do that quickly...

**Comment by impjdi** [comment]
commit 04a8d6e3f4eabae3b06cfdba942ec75da56e55d8 should improve logging.

**Comment by mtngld** [comment]
Thank you for the improved logging - 

```
10-06 14:17:17.732 19342 19342 E tflite  : TfLiteGpuDelegate Init: No shader implementation for const
```

There is definitely some issue with the shader implementation for constants, however I think this is not the root problem of my model and only arises from the toy example I provided (model composed of two constants)


I have edited my toy example to use Keras API, here are three different variants -

### Concatenate input layer and constant layer, BxHxWxC - Works Fine

```python
import tensorflow as tf

shape = (1, 1, 1, 10)
# Create a model using high-level tf.keras.* APIs
input = tf.keras.Input(shape=shape[1:], batch_size=1)
initializer = tf.keras.initializers.Constant(value=0)
const = initializer(shape=shape)
output = tf.keras.layers.concatenate([input, const])
model = tf.keras.Model(inputs=[input], outputs=output)
model.summary()
model.compile() # compile the model

# Convert the model.
converter = tf.lite.TFLiteConverter.from_keras_model(model)
tflite_model = converter.convert()

# Save the model.
with open('model.tflite', 'wb') as f:
  f.write(tflite_model)
```

### Concatenate two input layers BxC - Works fine
```python
import tensorflow as tf

shape = (1, 10)
# Create a model using high-level tf.keras.* APIs
input = tf.keras.Input(shape=shape[1:], batch_size=1)
input_2 = tf.keras.Input(shape=shape[1:], batch_size=1)
output = tf.keras.layers.concatenate([input, input_2])
model = tf.keras.Model(inputs=[input, input_2], outputs=output)
model.summary()
model.compile() # compile the model

# Convert the model.
converter = tf.lite.TFLiteConverter.from_keras_model(model)
tflite_model = converter.convert()

# Save the model.
with open('model.tflite', 'wb') as f:
  f.write(tflite_model)
```

### Concatenate input layer and constant layer, BxC - Fails

```python
import tensorflow as tf

shape = (1, 10)
# Create a model using high-level tf.keras.* APIs
input = tf.keras.Input(shape=shape[1:], batch_size=1)
initializer = tf.keras.initializers.Constant(value=0)
const = initializer(shape=shape)
output = tf.keras.layers.concatenate([input, const])
model = tf.keras.Model(inputs=[input], outputs=output)
model.summary()
model.compile() # compile the model

# Convert the model.
converter = tf.lite.TFLiteConverter.from_keras_model(model)
tflite_model = converter.convert()

# Save the model.
with open('model.tflite', 'wb') as f:
  f.write(tflite_model)
```

failed logcat:
```
10-06 14:49:30.819 23274 23274 I tflite_BenchmarkModelActivity: Running TensorFlow Lite benchmark with args: --graph=/data/local/tmp/model.tflite \     --num_threads=4 --use_gpu=true
10-06 14:49:30.827 23274 23274 I tflite  : Initialized TensorFlow Lite runtime.
10-06 14:49:30.827 23274 23274 I tflite  : Created TensorFlow Lite delegate for GPU.
10-06 14:49:30.828 23274 23274 E tflite  : TfLiteGpuDelegate Init: CONCATENATION: Expected a 4D tensor of shape BxHxWxC but got 1x10
10-06 14:49:30.828 23274 23274 I tflite  : Created 0 GPU delegate kernels.
10-06 14:49:30.828 23274 23274 E tflite  : TfLiteGpuDelegate Prepare: delegate is not initialized
10-06 14:49:30.828 23274 23274 E tflite  : Node number 1 (TfLiteGpuDelegateV2) failed to prepare.
10-06 14:49:30.828 23274 23274 E tflite  : Restored original execution plan after delegate application failure.
```

### Conclusion

The specific case where the input layer is concatenated with constant layer AND the dimensions are not BHWC fails, other use cases are fine.

**Comment by impjdi** [comment]
Thanks for the experiments.  So it's just the matter of `CONCAT` not being able to handle a const tensor as input tensor.  In that case, could you use `PAD` the input tensor with zeroes and expand the constant tensor with zeroes, so that you can achieve a `CONCAT` through a `PAD` followed by an `ADD`?

**Comment by mtngld** [comment]
The PAD workaround seems to be partially working 

However PAD by itself doesn't seem to support 2D tensors - 

```
10-06 23:18:51.113  9281  9281 I tflite  : Initialized TensorFlow Lite runtime.
10-06 23:18:51.115  9281  9281 I tflite  : Created TensorFlow Lite delegate for GPU.
10-06 23:18:51.117  9281  9281 E tflite  : Following operations are not supported by GPU delegate:
10-06 23:18:51.117  9281  9281 E tflite  : PAD: Invalid paddings tensor shape: expected 4x2 or 3x2, got 2x2
10-06 23:18:51.117  9281  9281 E tflite  : 81 operations will run on the GPU, and the remaining 38 operations will run on the CPU.
10-06 23:18:53.464  9281  9281 I tflite  : Initialized OpenCL-based API.
10-06 23:18:53.681  9281  9281 I tflite  : Created 1 GPU delegate kernels.
```

So at least it doesn't fail on GPU, but seems like some of the graph still runs on CPU.

**Comment by impjdi** [comment]
Thanks for trying.  That's annoying.  The only cheating I can think of is inserting the content of the constant tensor as the graph's input tensor.  Beyond this, I'm out of ideas...

We might have to support the constant input tensor case properly, but that requires a larger plumbing (it's not just updating the shader impl); we probably can't prioritize that at the moment.

**Comment by sushreebarsa** [comment]
Was able to replicate the issue with TF 2.5,please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/0601e2344a547c11e4afd1ca59deb8a9/untitled287.ipynb) ..Thanks!

**Comment by impjdi** [comment]
I cannot immediately work on this because I have other higher priorities.  Attaching model for now.
[concat_fail.tflite.zip](https://github.com/tensorflow/tensorflow/files/6687903/concat_fail.tflite.zip)

**Comment by sushreebarsa** [comment]
I am getting output in both [GPU](https://colab.research.google.com/drive/1EPjFqyfwLcsPdBo4Y6wamZnay2t1IcFr?resourcekey=0-6w6oN4cJJPuLMB_Nhr4qdg#scrollTo=6_Ygo8suaYQN) and[ CPU](https://colab.research.google.com/gist/sushreebarsa/0601e2344a547c11e4afd1ca59deb8a9/untitled287.ipynb) with TF v2.5 in Colab....Please find the gists and let us know if it is good to close ..Thanks !
